p4-16/spec/P4-16-spec.mdk

```diff
@@ -3220,6 +3222,20 @@ type. The result always has the same width as the left operand.
   - all result bits are zero when shifting left
   - all result bits are zero when shifting right a positive value
   - all result bits are one when shifting right a negative value
+- Extraction of a set of contiguous bits, also known as a slice, denoted by `[m:l]`,
+  where `m` and `l` must be positive integers
+  that are compile-time known values, and `m >= l`. The result is
+  a bit-string of width `m - l + 1`, including the bits numbered
+  from `l` (which becomes the least significant bit of the result) to `m` (the
+  most significant bit of the result) from the source operand. The conditions `0 <= l < W`
+  and `l <= m < W` are checked statically (where `W` is
+  the length of the source bit-string). Note that both endpoints of
+  the extraction are inclusive. The bounds are required to be
+  compile-time known values so that the result width can be computed
+  at compile time. Slices are also l-values, which means that P4 supports assigning to a slice: ` e[m:l] = x `.
+  The effect of this statement is to set bits `m` to `l` of `e` to the
+  bit-pattern represented by `x`, and leaves all other bits of `e`
+  unchanged.  A slice of a signed integer is treated like a signed integer as well.
```

Comment:
(Documenting and expanding on what I mentioned in the LDWG meeting on 3 Dec 2018.)

To me, bit-slicing is fundamentally an operation on bit strings, not integers. It seems strange, then, for the result of a bit slice to be signed. I can think of two cases where it makes sense to have a signed result:
1. the bit slice is being used as an r-value, and is truncating an `int<W>`, which is equivalent to casting to a shorter integer type; and
2. the bit slice includes the sign bit, which is equivalent to an arithmetic right shift, followed by a cast.

I'd argue that these use cases are an abuse of bit slicing (programmers should instead be doing the alternative equivalents mentioned above) and that bit slicing should always produce an unsigned result.

Reply:
Writing without my LDWG co-chair hat on, I endorse this perspective. It is simple and clean. And adding the required casts doesn't seem like it would clutter up programs in practical scenarios.

Reply:
I am not arguing _for_ the result of a bit-slice on `int<W>` types to be signed, but note that if you consider cases 1 and 2 to make sense to have signed results, 1 could be considered simply a "synonym" for `my_int[W-1:0]`, and 2 as a synonym for `my_int[width-1:lsb]`.  If you think of the 'extreme cases' as reasonable to have signed results, then I'd say it makes just as sense for all of the intermediate bit slices to have signed results, too.

Again, I am perfectly happy to simply just say that the result of a bit slice on a signed integer is an unsigned result always.  Bits are bits, and none of this wrangling over the types changes the bit patterns of the results at all.